beard/clustering/blocking.py

```diff
@@ -99,52 +158,38 @@ def _validate(self, X, blocks):
 
     def _fit(self, X, y, blocks):
         """Fit base clustering estimators on X."""
+        def _grouper(jobs, seq):
+            seq_len = len(seq)
+            group_size_by_job = (seq_len // jobs) * np.ones(jobs, dtype=np.int)
+            group_size_by_job[:seq_len % jobs] += 1
+            starts = [0] + np.cumsum(group_size_by_job).tolist()
+            return (seq[starts[i]: starts[i+1]] for i in range(0, jobs))
+
         self.labels_ = -np.ones(len(X), dtype=np.int)
         offset = 0
 
-        for b in np.unique(blocks):
-            # Select data from block
-            mask = (blocks == b)
-            X_mask = X[mask, :]
+        true_jobs = positive_n_jobs(self.n_jobs)
+        unique_blocks = np.unique(blocks)
 
-            if y is not None:
-                y_mask = y[mask]
-            else:
-                y_mask = None
-
-            if self.affinity == "precomputed":
-                X_mask = X_mask[:, mask]
-
-            # Fit a clusterer on the selected data
-            if len(X_mask) == 1:
-                clusterer = _SingleClustering()
-            elif self.fit_:
-                clusterer = clone(self.base_estimator)
-            elif self.partial_fit_:
-                if b in self.clusterers_:
-                    clusterer = self.clusterers_[b]
-                else:
-                    clusterer = clone(self.base_estimator)
-
-            if self.fit_ or not hasattr(clusterer, "partial_fit"):
-                try:
-                    clusterer.fit(X_mask, y=y_mask)
-                except TypeError:
-                    clusterer.fit(X_mask)
-            elif self.partial_fit_:
-                try:
-                    clusterer.partial_fit(X_mask, y=y_mask)
-                except TypeError:
-                    clusterer.partial_fit(X_mask)
-
-            self.clusterers_[b] = clusterer
-
-            # Save predictions
-            pred = np.array(clusterer.labels_)
-            mask_unknown = (pred == -1)
-            pred[~mask_unknown] += offset
-            self.labels_[mask] = pred
-            offset += np.max(clusterer.labels_) + 1
+        parallelizer = Parallel(n_jobs=self.n_jobs)
+        tasks_iterator = (delayed(_parallel_fit)(X, y, blocks, self.affinity,
+                                                 self.fit_,
+                                                 self.partial_fit_,
+                                                 self.base_estimator,
+                                                 self.clusterers_, b)
+                          for b in _grouper(true_jobs, unique_blocks))
+
+        result = parallelizer(tasks_iterator)
```

Comment:
Can you make the three lines above a single statement? `parallelizer` and `tasks_iterator` variables have no real use here.